### PR TITLE
MINIFICPP-2323 ListenTCP custom delimiter

### PR DESCRIPTION
**Pull request by martinzink**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: PROCESSORS.md [hunk 1]
```diff
@@ -1561,6 +1561,8 @@ In the list below, the names of required properties appear in bold. Any other pr
...
 | SSL Context Service           |               |                            | The Controller Service to use in order to obtain an SSL Context. If this property is set, messages will be received over a secure connection.                                                    |
 | Client Auth                   | NONE          | NONE<br/>WANT<br/>REQUIRED | The client authentication policy to use for the SSL Context. Only used if an SSL Context Service is provided.                                                                                    |
+| **Message Delimiter**         |               |                            | The delimiter is used to divide the stream into flowfiles.                                                                                                                                       |
+| **Consume delimiter**         | true          | true<br/>false             | If set to true then the delimiter won't be included in the resulting flowfiles.                                                                                                                  |
```
Comment (szaszm) [inline PROCESSORS.md:1565]: A bit more info.
```suggestion
| **Consume delimiter**         | true          | true<br/>false             | If set to true then the delimiter won't be included at the end of the resulting flowfiles.                                                                                                       |
```

File: PROCESSORS.md [hunk 2]
```diff
@@ -1561,6 +1561,8 @@ In the list below, the names of required properties appear in bold. Any other pr
 | **Max Size of Message Queue** | 10000         |                            | Maximum number of messages allowed to be buffered before processing them when the processor is triggered. If the buffer is full, the message is ignored. If set to zero the buffer is unlimited. |
 | SSL Context Service           |               |                            | The Controller Service to use in order to obtain an SSL Context. If this property is set, messages will be received over a secure connection.                                                    |
 | Client Auth                   | NONE          | NONE<br/>WANT<br/>REQUIRED | The client authentication policy to use for the SSL Context. Only used if an SSL Context Service is provided.                                                                                    |
+| **Message Delimiter**         |               |                            | The delimiter is used to divide the stream into flowfiles.                                                                                                                                       |
```
Comment (szaszm) [inline PROCESSORS.md:1564]: The default value of '\n' is not visible, probably it's just not escaped.

File: PROCESSORS.md [hunk 3]
```diff
@@ -1548,7 +1548,7 @@ In the list below, the names of required properties appear in bold. Any other pr
 
 ### Description
 
-Listens for incoming TCP connections and reads data from each connection using a line separator as the message demarcator. For each message the processor produces a single FlowFile.
+Listens for incoming TCP connections and reads data from each connection using a configurable message demarcator. For each message the processor produces a single FlowFile.
```
Comment (szaszm) [inline PROCESSORS.md:1551]: We should call it either delimiter or demarcator consistently.
```suggestion
Listens for incoming TCP connections and reads data from each connection using a configurable message delimiter. For each message the processor produces a single FlowFile.
```

File: PROCESSORS.md [hunk 4]
```diff
@@ -1561,6 +1561,8 @@ In the list below, the names of required properties appear in bold. Any other pr
...
 | SSL Context Service           |               |                            | The Controller Service to use in order to obtain an SSL Context. If this property is set, messages will be received over a secure connection.                                                    |
 | Client Auth                   | NONE          | NONE<br/>WANT<br/>REQUIRED | The client authentication policy to use for the SSL Context. Only used if an SSL Context Service is provided.                                                                                    |
+| **Message Delimiter**         | \n            |                            | The delimiter is used to divide the stream into flowfiles.                                                                                                                                       |
+| **Consume delimiter**         | true          | true<br/>false             | If set to true then the delimiter won't be included at the end of the resulting flowfiles.                                                                                                       |
```
Comment (szaszm) [inline PROCESSORS.md:1565]: Most properties have capital first letters in each word.
```suggestion
| **Consume Delimiter**         | true          | true<br/>false             | If set to true then the delimiter won't be included at the end of the resulting flowfiles.                                                                                                       |
```

File: extensions/standard-processors/processors/ListenTCP.h [hunk 5]
```diff
@@ -36,10 +36,10 @@ namespace org::apache::nifi::minifi::processors {
 class ListenTCP : public NetworkListenerProcessor {
  public:
   explicit ListenTCP(std::string_view name, const utils::Identifier& uuid = {})
-    : NetworkListenerProcessor(name, uuid, core::logging::LoggerFactory<ListenTCP>::getLogger(uuid)) {
+      : NetworkListenerProcessor(name, uuid, core::logging::LoggerFactory<ListenTCP>::getLogger(uuid)) {
   }
 
-  EXTENSIONAPI static constexpr const char* Description = "Listens for incoming TCP connections and reads data from each connection using a line separator as the message demarcator. "
+  EXTENSIONAPI static constexpr const char* Description = "Listens for incoming TCP connections and reads data from each connection using a configurable message demarcator. "
```
Comment (szaszm) [inline extensions/standard-processors/processors/ListenTCP.h:42]:
```suggestion
  EXTENSIONAPI static constexpr const char* Description = "Listens for incoming TCP connections and reads data from each connection using a configurable message delimiter. "
```

File: extensions/standard-processors/processors/ListenTCP.h [hunk 6]
```diff
@@ -69,15 +69,29 @@ class ListenTCP : public NetworkListenerProcessor {
...
       .withAllowedValues(magic_enum::enum_names<utils::net::ClientAuthOption>())
       .build();
-  EXTENSIONAPI static constexpr auto Properties = std::array<core::PropertyReference, 5>{
+  EXTENSIONAPI static constexpr auto MessageDelimiter = core::PropertyDefinitionBuilder<>::createProperty("Message Delimiter")
+      .withDescription("The delimiter is used to divide the stream into flowfiles.")
+      .isRequired(true)
+      .withDefaultValue("\n")
+      .supportsExpressionLanguage(false)
+      .build();
+  EXTENSIONAPI static constexpr auto ConsumeDelimiter = core::PropertyDefinitionBuilder<>::createProperty("Consume delimiter")
```
Comment (szaszm) [inline extensions/standard-processors/processors/ListenTCP.h:78]:
```suggestion
  EXTENSIONAPI static constexpr auto ConsumeDelimiter = core::PropertyDefinitionBuilder<>::createProperty("Consume Delimiter")
```

**Comment by lordgamez** [comment]
I think the Jira ticket for this issue is duplicate of this: https://issues.apache.org/jira/browse/MINIFICPP-2310

**Review by szaszm — APPROVED** [review]

File: extensions/standard-processors/processors/ListenTCP.cpp [hunk 7]
```diff
@@ -28,7 +28,13 @@ void ListenTCP::initialize() {
 }
 
 void ListenTCP::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
-  startTcpServer(context, SSLContextService, ClientAuth);
+  auto delimiter_str = context.getProperty(MessageDelimiter).value_or("\n");
+  if (delimiter_str.empty()) {
+    logger_->log_warn("{} cannot be an empty string", MessageDelimiter.name);
```
Comment (fgerlits) [inline extensions/standard-processors/processors/ListenTCP.cpp:33]: I think we should tell the user what is happening:
```suggestion
    logger_->log_warn("{} cannot be an empty string, using \\n as the delimiter", MessageDelimiter.name);
```

Comment (martinzink) [reply extensions/standard-processors/processors/ListenTCP.cpp]: good idea, ive included this in https://github.com/apache/nifi-minifi-cpp/pull/1753/commits/0f488acc7ebffe1ac35e1a0924c96b63a15f5eba

File: extensions/standard-processors/processors/ListenTCP.cpp [hunk 8]
```diff
@@ -28,7 +28,13 @@ void ListenTCP::initialize() {
...
+    delimiter_str = "\n";
+  }
```
[the remark this answers is not in the document]
Comment (martinzink) [reply extensions/standard-processors/processors/ListenTCP.cpp]: Good idea, but I didnt want to limit this processor to a single char so I've added a new function `replaceEscapedCharacters` in https://github.com/apache/nifi-minifi-cpp/pull/1753/commits/0f488acc7ebffe1ac35e1a0924c96b63a15f5eba

File: libminifi/src/utils/StringUtils.cpp [hunk 9]
```diff
@@ -538,6 +538,56 @@ nonstd::expected<std::optional<char>, ParseError> parseCharacter(std::string_vie
...
+      case 'f':
+        result << '\f';  // Form Feed
+        ++i;
+        break;
+      case 'r':
+        result << '\r';  // Carriage Return
+        ++i;
+        break;
+      default:
+        result << '\\';
```
Comment (fgerlits) [inline libminifi/src/utils/StringUtils.cpp:584]: This is an interesting question: should `"\s"` expand to `"s"` or `"\s"`?  The old `parseCharacter` function does the first, here you do the second. I think it's a bit confusing that they are different, but I can be persuaded either way.

File: libminifi/test/unit/StringUtilsTests.cpp [hunk 10]
```diff
@@ -615,6 +615,16 @@ TEST_CASE("string::parseCharacter tests") {
   CHECK(string::parseCharacter("") == std::nullopt);
 }
 
+TEST_CASE("string::replaceEscapedCharacters tests") {
+  CHECK(string::replaceEscapedCharacters("a") == "a");
+  CHECK(string::replaceEscapedCharacters(R"(\n)") == "\n");
+  CHECK(string::replaceEscapedCharacters(R"(\t)") == "\t");
+  CHECK(string::replaceEscapedCharacters(R"(\r)") == "\r");
+  CHECK(string::replaceEscapedCharacters(R"(\\s)") == "\\\\s");
```
Comment (fgerlits) [inline libminifi/test/unit/StringUtilsTests.cpp:623]: I would add
```
CHECK(string::replaceEscapedCharacters(R"(\s)") == "\\s");
```
(or
```
CHECK(string::replaceEscapedCharacters(R"(\s)") == "s");
```
if we decide to change it)

**Comment by szaszm** [comment]
Do you want to fix the clang-tidy warnings?

**Comment by martinzink** [comment]
> Do you want to fix the clang-tidy warnings?

These wouldnt be trivial changes so Id rather not do these here from https://issues.apache.org/jira/browse/MINIFICPP-2247 

> libminifi/src/utils/StringUtils.cpp contains a couple of functions that use c style arrays which should be replaced by std::array, the changes are not trivial so we probably want to create a separate PR instead of doing in whichever PR that touches the file

File: libminifi/test/unit/StringUtilsTests.cpp (hunk 10 above)
Comment (martinzink) [reply libminifi/test/unit/StringUtilsTests.cpp]: I've changed these based on your suggestion: https://github.com/apache/nifi-minifi-cpp/pull/1753/commits/3b6a8c40b0f991513b71e3c2a53bef7b6a088acc

**Review by fgerlits — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**